File: frontend/src/api/jobs.ts
import { api } from './client'

export interface TaskStatus {
  key: string
  life_cycle: string
  result?: string
  display_state: string
  font_color: string
  fill_color: string
  depends_on: string[]
}

export interface RunStatus {
  run_id: number
  life_cycle: string
  result?: string
  message: string
  is_running: boolean
  is_done: boolean
  elapsed_seconds?: number
  tasks: TaskStatus[]
}

export interface WorkflowConfig {
  team: string
  job_id: string
  sheet_id: string
  job_url: string
  sheet_url: string
}

export interface JobDetail {
  job_id: number
  name: string
  last_run_at?: string
  active_run_id?: number
  parameters: { name: string; default: string }[]
  tasks: {
    key: string
    type: string
    path: string
    params: string
    depends_on: string[]
    cluster: string
  }[]
}

export async function getWorkflowConfig(team: string): Promise<WorkflowConfig | null> {
  const res = await api.get<WorkflowConfig | null>('/jobs/workflow-config', { params: { team } })
  return res.data
}

export async function getJobDetails(jobId: number): Promise<JobDetail> {
  const res = await api.get<JobDetail>(`/jobs/${jobId}/details`)
  return res.data
}

export async function triggerJob(jobId: number): Promise<{ run_id: number; run_url: string }> {
  const res = await api.post<{ run_id: number; run_url: string }>(`/jobs/${jobId}/run`)
  return res.data
}

export async function cancelRun(runId: number): Promise<void> {
  await api.post(`/jobs/runs/${runId}/cancel`)
}

export async function getRunStatus(runId: number): Promise<RunStatus> {
  const res = await api.get<RunStatus>(`/jobs/runs/${runId}/status`)
  return res.data
}

File: backend/routers/jobs.py
import asyncio
import json
from datetime import datetime

from fastapi import APIRouter, HTTPException
from sse_starlette.sse import EventSourceResponse

from backend.config import WORKFLOW_CONFIG
from backend.models.job import JobDetail, RunStatus, TaskStatus, WorkflowConfig
from backend.services.job_monitor import cancel_run, get_job_details, get_run_status, trigger_job

router = APIRouter(prefix="/jobs", tags=["jobs"])

_POLL_INTERVAL_S = 15


@router.get("/workflow-config")
async def get_workflow_config(team: str) -> WorkflowConfig | None:
    cfg = WORKFLOW_CONFIG.get(team)
    if not cfg:
        return None
    return WorkflowConfig(team=team, **cfg)


@router.get("/{job_id}/details", response_model=JobDetail)
async def job_details(job_id: int):
    try:
        return get_job_details(job_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{job_id}/run")
async def run_job(job_id: int):
    try:
        return trigger_job(job_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/runs/{run_id}/cancel")
async def cancel_job_run(run_id: int):
    try:
        cancel_run(run_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    return {"ok": True}


@router.get("/runs/{run_id}/status", response_model=RunStatus)
async def run_status(run_id: int):
    try:
        data = get_run_status(run_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    return RunStatus(
        **{k: v for k, v in data.items() if k != "tasks"},
        tasks=[TaskStatus(**t) for t in data["tasks"]],
    )


@router.get("/runs/{run_id}/stream")
async def stream_run_status(run_id: int):
    start = datetime.now()

    async def event_generator():
        while True:
            try:
                data = get_run_status(run_id)
            except Exception as e:
                yield {"event": "error", "data": json.dumps({"detail": str(e)})}
                return

            elapsed = int((datetime.now() - start).total_seconds())
            status = RunStatus(
                **{k: v for k, v in data.items() if k != "tasks"},
                elapsed_seconds=elapsed,
                tasks=[TaskStatus(**t) for t in data["tasks"]],
            )
            payload = status.model_dump_json()

            if data["is_done"]:
                yield {"event": "done", "data": payload}
                return
            else:
                yield {"event": "status", "data": payload}

            await asyncio.sleep(_POLL_INTERVAL_S)

    return EventSourceResponse(event_generator())
